### PR TITLE
Render Govspeak preview instead of plain Markdown

### PR DESCRIPTION
**Pull request by benthorner**
https://trello.com/c/77pWuPzi/170-render-govspeak-server-side

This commit changes the markdown-editor component to do an async query
to a new protected API that renders the passed Govspeak as HTML.

This requires the component to have knowledge of the API path via a new
data attribute, which is injected deep inside the component hierarchy.
If we were to move this component outside of the app, it ought to have a
default Govspeak rendering service it can use.

Rendering the markdown asynchronously has two error cases: network error
and application error. There's no evidence that an application error is
possible for this API, so we now only handle the network error in JS.

I also considered renaming markdown-* to govspeak-*, but it appears we
refer to 'govspeak' and 'markdown' interchangeably, so left as-is.

### CODE REVIEW

File: app/javascript/components/markdown-editor.js [hunk 1]
```diff
@@ -38,32 +37,38 @@ MarkdownEditor.prototype.handlePreviewButton = function (event) {
...
-    // Open preview links in a new tab
-    this.setTargetBlank(this.$preview)
-  } else {
+  if (!text) {
     $preview.innerHTML = 'Nothing to preview'
+    return
   }
+
+  var path = $module.getAttribute('data-govspeak-path')
+  var url = new URL(document.location.origin + path)
```
Comment (alex-ju) [inline app/javascript/components/markdown-editor.js]: `URL` and `searchParams` are not supported in any version of Internet Explorer

File: app/javascript/components/markdown-editor.js [hunk 2]
```diff
@@ -38,32 +37,38 @@ MarkdownEditor.prototype.handlePreviewButton = function (event) {
...
+  url.searchParams.append('govspeak', text)
+
+  // Render markdown
+  window.fetch(url, { credentials: 'include' })
```
[the remark this answers is not in the document]
Comment (alex-ju) [reply app/javascript/components/markdown-editor.js]: `fetch` is not supported by IE8 either, but I'm aware it's being used for validating the slug too and we're planning to use a polyfill for this.

[the remark this answers is not in the document]
Comment (benthorner) [reply app/javascript/components/markdown-editor.js]: Good spot @tijmenb. There's no official timeout for fetch, so I'll have a look at an alternative as we should have one for this and the URL preview.

File: app/javascript/components/markdown-editor.js [hunk 3]
```diff
@@ -38,32 +37,38 @@ MarkdownEditor.prototype.handlePreviewButton = function (event) {
...
-    this.setTargetBlank(this.$preview)
-  } else {
+  if (!text) {
     $preview.innerHTML = 'Nothing to preview'
+    return
   }
+
+  var path = $module.getAttribute('data-govspeak-path')
+  var url = new URL(document.location.origin + path)
+  url.searchParams.append('govspeak', text)
```
[the remark this answers is not in the document]
Comment (benthorner) [reply app/javascript/components/markdown-editor.js]: Good point - I had this thought too and forgot about it. POST is definitely the way to go...

File: config/routes.rb [hunk 4]
```diff
@@ -26,6 +26,8 @@
 
   get "/documentation" => "documentation#index"
 
+  get "/govspeak" => "govspeak#to_html"
```
[the remark this answers is not in the document]
Comment (benthorner) [reply config/routes.rb]: Fine with me. I'll also extract out the Govspeak rendering line from the payload service, so we're generating govspeak in the same way each time.

File: app/javascript/components/markdown-editor.js (hunk 1 above)
Comment (benthorner) [reply app/javascript/components/markdown-editor.js]: I'll go for Polyfill - die die XMLHttpRequest!

**Comment by benthorner** [comment]
Comments should all be addressed now - @tijmenb the timeout is (arbitrarily) 5 secs :-).

**Pull request merged**